Initial RFE for SBOM SPDX support (#11)

javihernandez opened the pull request.

javihernandez committed:
Initial RFE for SBOM SPDX support

https://github.com/AlmaLinux/alma-sbom/issues/11

javihernandez commented on SBOM/spdx.md:
```diff
@@ -0,0 +1,216 @@
+
+In addition to that, we only add the data that is specific to SPDX specification. At the time of writing this RFE, the current version of the SPDX specification is [v2.3](https://spdx.github.io/spdx-spec/v2.3/).
+
+**TODO**: Describe SPDX-specific data (or fields)
```
@m10k do you mind taking care of filling this out?

javihernandez commented on SBOM/spdx.md:
```diff
@@ -0,0 +1,216 @@
+    "dataLicense": "CC0-1.0",
+    "name": "bash-5.1.8-6.el9_1",
+    "spdxVersion": "SPDX-2.3",
+    "documentNamespace": "https://almalinux.org/spdx",
```
The URI does not have to be accessible. It is only intended to provide a unique ID. In many cases, the URI will point to a Web accessible document, but this should not be assumed to be the case.

If we follow this pattern: https://[CreatorWebsite]/[pathToSpdx]/[DocumentName]-[UUID]

Should we end up with something similar to: https://security.almalinux.org/spdx/bash-5.1.8-6.el9_1-424bb7a2-e4d7-472b-8260-23ec0feb7f07 ?

m10k replied:
This sounds good to me. This way all SBOMs have a unique id/namespace and we can make them downloadable at a later time (if it becomes necessary).

javihernandez replied:
Good, can you please incorporate such change into the PR on `alma-sbom`?

m10k replied:
Yep, I just added a commit that addresses this issue

javihernandez commented on SBOM/spdx.md:
```diff
@@ -0,0 +1,216 @@
+                    "comment": "almalinux:albs:build:source:gitCommitCasHash=70c5a02df7074eada535aa35643f019574f7005ff345fd6f482eca797ae7c1d8"
+                }
+            ],
+            "builtDate": "2023-01-24T09:44:30.238513+00:00Z",
```
`builtDate` should follow `YYYY-MM-DDThh:mm:ssZ` format

m10k replied:
You're right, I'll fix my PR. I assumed the format was okay because the validator didn't complain about it. I'll check if this is a known issue with spdx-tools and report it upstream if necessary.

For the record:
https://spdx.github.io/spdx-spec/v2.3/package-information/#726-built-date

javihernandez replied:
Had the same impression here, as you said, maybe double check if this is something going wrong on `spdx-tools` side. Meanwhile, yes, please do fix your PR to `alma-sbom`

m10k replied:
There's definitely something wrong with spdx-tools. Even assigning a `datetime.now()` causes the timestamp to be in incorrect format (it includes milliseconds). I have to check if I can enforce the right format from outside spdx-tools.

m10k replied:
I created an issue upstream [1] and one of the developer fixed it just a few minutes ago [2].

[1] https://github.com/spdx/tools-python/issues/718
[2] https://github.com/spdx/tools-python/pull/719

javihernandez commented on SBOM/spdx.md:
```diff
@@ -0,0 +1,216 @@
+                    "checksumValue": "1bfcf5bf3187ff21db8109d48a0198b92341a9124f63db82cf501d1968ba1374"
+                }
+            ],
+            "downloadLocation": "pkg:rpm/almalinux/bash@5.1.8-6.el9_1?arch=x86_64&upstream=bash-5.1.8-6.el9_1.src.rpm",
```
We need to figure out whether this is mandatory, as I see:

If the reference is BINARY, its relationship to `PackageDownloadLocation` might need to be explained. If the reference is SOURCE, its relationship to `PackageDownloadLocation` and `SourceInformation` might need to be explained.
We're already providing a purl, isn't that already enough?

If need to include it:
 * shouldn't this be `packageDownloadLocation`?
 * do we need to come up with a solution for providing the right `downloadLocation`? what we have when generating the sbom is the artifact that came off the ALBS. Or, are we okay with just providing the info from that one?

m10k replied:
I should have read the specification more carefully.. apparently, we can just set this to `NONE` (or `NOASSERTION`) and that's it.

The specification also suggests linking to the source tree, which I would prefer. The necessary information is contained in `almalinux:albs:build:source:gitURL` and `almalinux:albs:build:source:gitCommit`. In this example, we'd get the URL `git+https://git.almalinux.org/rpms/bash.git@1a3ebc6398c31698863ba9b31a9f5effdb425c4f`.

Regarding the field name, I think `packageDownloadLocation` is the name in tag-value format. In the rdf examples, the specification uses the following notation.

```xml
<Package rdf:about="...">
    <downloadLocation>URL</downloadLocation>
</Package>
```

Source:
https://spdx.github.io/spdx-spec/v2.3/package-information/#77-package-download-location-field

javihernandez replied:
Okay, that makes sense, let's just set it to `NONE` for now.
And yes, got confused about `downloadLocation` vs `packageDownloadLocation` when reading the spec back and forth :sweat_smile:

m10k replied:
I just added a commit that sets the `downloadLocation` to `NONE` until we have something better.

m10k commented on SBOM/spdx.md:
```diff
@@ -0,0 +1,216 @@
+* XML
+* Tag/Value
+* YAML
+
```
There is also support for rdf in spdx-tools, but I couldn't get it to play nice with my python version. I didn't try to find out if the problem was in my environment or in spdx-tools, but when the next stable spdx-tools is published, we should look at this again.

javihernandez replied:
Good to know, we can add this into the RFE and then deal with the problem ;)

m10k committed:
SBOM/spdx: Add information about SPDX-specific fields

This commit adds explanations of SPDX-specific fields to the SPDX
SBOM RFE.

m10k committed:
SBOM/spdx: Update author information

The author information in the SPDX SBOM RFE doesn't have my email
address.

This commit adds my email address to the author information of the
SPDX SBOM RFE.

m10k committed:
SBOM/spdx: Add note about RDF support

This commit adds a short note about RDF support in spdx-tools and
its current state in alma-sbom.

armintaenzertng commented:
If you need any help in adopting the spdx-tools in your project, please feel free to ask! :)
We are planning to have an RC up soon.

javihernandez commented:
> If you need any help in adopting the spdx-tools in your project, please feel free to ask! :)
> We are planning to have an RC up soon.

Many thanks!! I guess that for now we're ok. I believe that @m10k is having issues when generating RDF-formatted SBOMs, we'll definitely come back to you if he can't sort that out.

In regard of having a new release of spdx-tools, that'd be great, specially because we'd like to use a proper release rather than current `spdx-tools/master` git branch.

javihernandez commented on SBOM/spdx.md:
```diff
@@ -0,0 +1,240 @@
+
+#### SPDX SBOM record of a build
+
+**TBD**
```
@m10k can you also provide an example of a SBOM record for a build?

m10k replied:
I just pushed a commit that includes a build SBOM. I also updated the other SBOM in the RFE, since the format has changed quite a bit with the recent changes to alma-sbom.

m10k committed:
SBOM/spdx: Add example SBOM for a build record

This commit adds an example of an SBOM for a build, containing
multiple packages.

m10k committed:
SBOM/spdx: Update example of SPDX SBOM for an individual package

The SPDX SBOM for an individual package, that is shown in the RFE is
outdated.

This commit updates the RFE so the example shows an SBOM generated
by the latest version of alma-sbom.

anfimovdm commented on SBOM/spdx.md:
````diff
@@ -0,0 +1,1649 @@
+
+#### SPDX SBOM record of a package
+
+```javascript
````
I guess it should be `json` instead, isn't it?

m10k replied:
You're right. I just changed it and updated the branch.

javihernandez replied:
For the record, I initially set that to `javascript` because `json` didn't play nice when there were comments inside the block. Now that comments are gone, yes, `json` is the way to go. Thanks @anfimovdm and @m10k

m10k committed:
SBOM/spdx: Change syntax highlighting of SBOM examples to json

The SBOM examples are displayed with javascript syntax highlighting,
which is not the most appropriate choice for json data.

This commit changes the syntax highlighting of the two SBOM examples
to json mode.

javihernandez committed:
SBOM/spdx: Added link to SPDX RFE into README.md

javihernandez committed:
SBOM/spdx: Shortened build SBOM record
